### PR TITLE
28450 exclude monitors from plot

### LINKED ISSUE #28450: Exclude monitors in plots
Monitors should be excluded from:

- [x] color fill plots
- [x] bin plots
- [ ] slice viewer ?
- [x] instrument viewer ?

The presence of monitor explodes the scale of the plot making detector data practically invisible.

### PR DESCRIPTION
**Pull request by MathieuTillet**
**Description of work.**
When attempting to use either bin or colorfill plot, the monitors are not shown, thus not completely dwarfing the relevant data.
  
**To test:**
- Open `ExternalData/Testing/Data/UnitTest/ILL/PANTHER/001723.nxs`
- Right-click Plot/Colorfill and check that there is viewable data, even when the color bar is linear, and that it is not to long to appear (if possible, check by comparing to the older version, which should not show anything interesting in linear color bar mode).
- Open `ExternalData/Testing/Data/UnitTest/ILL/D11/010585.nxs`
- Right-click Plot/Bin and check that the data is not dwarfed by the monitor value, and that the plot is not too slow.
- Open `ExternalData/Testing/Data/UnitTest/HET15869.raw`. Plot the colorfill, and check that the first lines are white. The scale should go up to 1e+8.

Partially fixes #28450 

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by NickDraper — CHANGES_REQUESTED** [review]
So this does not alter the majority of TOF workspaces as they use another function in that file get_matrix_2d_ragged, could you add this into that Fn as well, as it needs to work in a consistent way as far as our users are concerned.

You can use HET15869 as a test file (it is in our unit test data)

Thanks

**Comment by NickDraper** [comment]
@MathieuTillet Have you got a chance to fix this so I can re-review it?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 1]
```diff
@@ -342,6 +342,41 @@ def get_spectrum(workspace, wkspIndex, normalize_by_bin_width, withDy=False, wit
...
+    spectrum_info = workspace.spectrumInfo()
+    monitors_indices = [i for i in range(total_range) if spectrum_info.hasDetectors(i) and spectrum_info.isMonitor(i)]
+    monitor_count = len(monitors_indices)
+
+    # If possible, ie the detectors' indices are continuous, we return a range.
+    # If not, we return a numpy array
+    range_start = -1
+    range_end = total_range
+    is_range = True
+    for index, i in enumerate(monitors_indices):
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:363]: Can we avoid one character variable names, e.g. i here which can't be index as there is another index variable, so it needs a better name.  is it monitor_index?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 2]
```diff
@@ -342,6 +342,41 @@ def get_spectrum(workspace, wkspIndex, normalize_by_bin_width, withDy=False, wit
...
+    """
+    Find the detectors' spectrum indices
+    (ie every detector which is not a monitor)
+
+    :param workspace: a Workspace2D or an EventWorkspace
+    :return : the detectors' spectrum indices as a range if possible, else as a numpy array
+    """
+    total_range = workspace.getNumberHistograms()
+    spectrum_info = workspace.spectrumInfo()
+    monitors_indices = [i for i in range(total_range) if spectrum_info.hasDetectors(i) and spectrum_info.isMonitor(i)]
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:355]: I think I should be index here

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 3]
```diff
@@ -351,14 +386,13 @@ def get_bins(workspace, wkspIndex, withDy=False):
     :param withDy: if True, it will return the error in the "counts", otherwise None
 
     """
-    num_hist = workspace.getNumberHistograms()
-    x = range(0, num_hist)
+    x = get_detectors_indices(workspace)
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:389]: So x should be the detector_indices? could it be called that?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 4]
```diff
@@ -351,14 +386,13 @@ def get_bins(workspace, wkspIndex, withDy=False):
...
     y = []
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:390]: if y is just the y values then I'm kind of ok with that, but would prefer y_values as it indicates that it is a list.

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 5]
```diff
@@ -351,14 +386,13 @@ def get_bins(workspace, wkspIndex, withDy=False):
...
     dy = [] if withDy else None
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:391]: as above

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 6]
```diff
@@ -423,27 +457,38 @@ def get_matrix_2d_ragged(workspace, normalize_by_bin_width, histogram2D=False, t
     delta = np.finfo(np.float64).max
     min_value = np.finfo(np.float64).max
     max_value = np.finfo(np.float64).min
+    try:
+        sp_info = workspace.spectrumInfo()
+    except:
+        sp_info = None
+
     for i in range(num_hist):
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:465]: i is worskpace_index?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 7]
```diff
@@ -423,27 +457,38 @@ def get_matrix_2d_ragged(workspace, normalize_by_bin_width, histogram2D=False, t
...
+            min_value = min(min_value, xtmp.min())
+            max_value = max(max_value, xtmp.max())
+            diff = xtmp[1:] - xtmp[:-1]
+            delta = min(delta, diff.min())
     num_edges = int(np.ceil((max_value - min_value)/delta)) + 1
     x_centers = np.linspace(min_value, max_value, num=num_edges)
     y = mantid.plots.datafunctions.boundaries_from_points(workspace.getAxis(1).extractValues())
     z = np.empty([num_hist, num_edges], dtype=np.float64)
+
     for i in range(num_hist):
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:483]: workspace_index?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 8]
```diff
@@ -423,27 +457,38 @@ def get_matrix_2d_ragged(workspace, normalize_by_bin_width, histogram2D=False, t
...
         centers, ztmp, _, _ = mantid.plots.datafunctions.get_spectrum(
             workspace, i, normalize_by_bin_width=normalize_by_bin_width, withDy=False, withDx=False)
         f = interp1d(centers, ztmp, kind='nearest', bounds_error=False, fill_value=np.nan)
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:486]: what is f?

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 9]
```diff
@@ -423,27 +457,38 @@ def get_matrix_2d_ragged(workspace, normalize_by_bin_width, histogram2D=False, t
...
-        z[i] = f(x_centers)
+        if not (sp_info and sp_info.hasDetectors(i) and sp_info.isMonitor(i)):
+            z[i] = f(x_centers)
```
Comment (NickDraper) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:488]: and what is z?

**Review by NickDraper — CHANGES_REQUESTED** [review]
So this works nicely now, for regular and non-regular binned workspaces, I tested quite a few files and it worked well.  The only times it didn't are because the monitors are not marked correctly, I'll follow that up separately.

There are a few single value variable names that would be better renamed to something that describes the purpose of the variable better to make it easier for others to understand the code.

I'd also like to see unit tests for get_detectors_indices and get_bins added into dataFunctionsTest.py.

Good work.

File: Framework/PythonInterface/mantid/plots/datafunctions.py (hunk 7 above)
Comment (MathieuTillet) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:483]: Ok, I will change all of those

File: Framework/PythonInterface/mantid/plots/datafunctions.py (hunk 8 above)
Comment (MathieuTillet) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:486]: if I understand correctly, f is the interpolating function used to make sure every histogram has the same width in the end

**Comment by MathieuTillet** [comment]
It should be better now

**Review by NickDraper — APPROVED** [review]
Thanks for the changes, it is much easier to read now.  Good work 😄

**Pull request merged**